src/huggingface_hub/hf_api.py

```diff
@@ -1805,70 +1878,82 @@ def upload_file(
         "https://huggingface.co/username/my-model/blob/main/remote/file/path.h5"
         ```
         """
-        if repo_type not in REPO_TYPES:
-            raise ValueError(f"Invalid repo type, must be one of {REPO_TYPES}")
-
-        try:
-            token, name = self._validate_or_retrieve_token(
-                token, function_name="upload_file"
-            )
-        except ValueError:  # if token is invalid or organization token
-            if self._is_valid_token(path_or_fileobj):
-                warnings.warn(
-                    "`upload_file` now takes `token` as an optional positional"
-                    " argument. Be sure to adapt your code!",
-                    FutureWarning,
-                )
-                token, path_or_fileobj, path_in_repo, repo_id = (
-                    path_or_fileobj,
-                    path_in_repo,
-                    repo_id,
-                    token,
-                )
-            else:
-                raise ValueError("Invalid token passed!")
+        commit_summary = (
+            commit_summary
+            if commit_summary is not None
+            else f"Upload {path_in_repo} with huggingface_hub"
+        )
+        operation = CommitOperationAdd(
+            path_or_fileobj=path_or_fileobj,
+            path_in_repo=path_in_repo,
+        )
 
-        # Validate path_or_fileobj
-        if isinstance(path_or_fileobj, str):
-            path_or_fileobj = os.path.normpath(os.path.expanduser(path_or_fileobj))
-            if not os.path.isfile(path_or_fileobj):
-                raise ValueError(f"Provided path: '{path_or_fileobj}' is not a file")
-        elif not isinstance(path_or_fileobj, (RawIOBase, BufferedIOBase, bytes)):
-            # ^^ Test from: https://stackoverflow.com/questions/44584829/how-to-determine-if-file-is-opened-in-binary-or-text-mode
-            raise ValueError(
-                "path_or_fileobj must be either an instance of str or BinaryIO. If you"
-                " passed a fileobj, make sure you've opened the file in binary mode."
+        pr_url = self.create_commit(
+            repo_id=repo_id,
+            repo_type=repo_type,
+            operations=[operation],
+            commit_summary=commit_summary,
+            commit_description=commit_description,
+            token=token,
+            revision=revision,
+            create_pr=create_pr,
+        )
+        return (
+            pr_url
+            if pr_url is not None
+            else hf_hub_url(
+                repo_id, path_in_repo, revision=revision, repo_type=repo_type
             )
```

Comment:
maybe the server should return the commit URL (or the blob URL?) in the JSON if not in PR mode?

and it's consistent with the old endpoint:

[internal link] 
https://github.com/huggingface/moon-landing/blob/cb42e3de2011d92905a15bf584f3a6906ca52f18/server/app/gitViewerRoutes.ts#L1309-L1312

Bonus is that then you don't need to move `hf_hub_url` which would keep this diff smaller =)

Reply:
(my opinion is that we can change the semantic of the function to return the commit URL instead of the blob URL)